src/command.rs

```diff
@@ -424,34 +425,533 @@ impl <U:WebDriverExtensionRoute> ToJson for WebDriverMessage<U> {
+                    format!("{} was not a valid url", name)))));
+                if url.username() != "" ||
+                    url.password() != None ||
+                    url.path() != "/" ||
+                    url.query() != None ||
+                    url.fragment() != None {
+                        return Err(WebDriverError::new(
+                            ErrorStatus::InvalidArgument,
+                            format!("{} was not of the form host[:port]", name)));
+                    }
+                let mut port_key = String::with_capacity(name.len() + 4);
+                port_key.push_str(name);
+                port_key.push_str("Port");
+                if url.port() != None &&
+                    obj.contains_key(&*port_key) {
+                        return Err(WebDriverError::new(
+                                    ErrorStatus::InvalidArgument,
+                                    format!("{} supplied with a port as well as {}",
+                                            name, port_key)));
+                    }
+            },
+            None => return Err(WebDriverError::new(
+                ErrorStatus::InvalidArgument,
+                format!("{} was not a string", name)
+            ))
+        }
+        Ok(())
+    }
+
+    fn validate_port(&self, name: &str, value: &Json) -> WebDriverResult<()> {
+        match value {
+            &Json::I64(x) => {
+                if x < 0 || x > 2^16 - 1 {
+                    return Err(WebDriverError::new(
+                        ErrorStatus::InvalidArgument,
+                        format!("{} is out of range", name)))
+                }
+            }
+            _ => return Err(WebDriverError::new(
+                ErrorStatus::InvalidArgument,
+                format!("{} was not an integer", name)))
+        }
+        Ok(())
+    }
+
+    fn validate_timeouts(&self, value: &Json) -> WebDriverResult<()> {
+        let obj = try_opt!(value.as_object(),
+                           ErrorStatus::InvalidArgument,
+                           "timeouts capability was not an object");
+        for (key, value) in obj.iter() {
+            match &**key {
+                x @ "script" |
+                x @ "pageLoad" |
+                x @ "implicit" => {
+                    let timeout = try_opt!(value.as_i64(),
+                                           ErrorStatus::InvalidArgument,
+                                           format!("{} timeouts value was not an integer", x));
+                    if timeout < 0 {
+                        return Err(WebDriverError::new(ErrorStatus::InvalidArgument,
+                                                       format!("{} timeouts value was negative", x)))
+                    }
+                },
+                x => return Err(WebDriverError::new(ErrorStatus::InvalidArgument,
+                                                    format!("{} was not a valid timeouts capability", x)))
```

Comment:
Again, I think it will have unfortunate consequences of checking additional unknown keys. I think we should ignore them.

For example, Selenium currently sends "`page load`", which I think would cause an error to be returned here. For compatibility reasons we should ignore it.